In [10]:
# imports 
import numpy as np
import pandas as pd
import os
from ydata_profiling import ProfileReport

# base bank dataset

In [11]:
df = pd.read_csv("../dataset/bank/bank-full.csv", delimiter=';')
df

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [12]:
profile = ProfileReport(df, title="Profiling Base Bank dataset Report", correlations={
        "auto": {"calculate": False},
        "pearson": {"calculate": False},
        "spearman": {"calculate": False},
        "kendall": {"calculate": False},
        "phi_k": {"calculate": True},
        "cramers": {"calculate": True},
    })
os.makedirs("./../mount/eda_profile_reports", exist_ok=True)
profile.to_file("./../mount/eda_profile_reports/base_bank_dataset.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### Notes

#### Null values

- There are no case of Null value entries. 
- However some of the fields are having some unknown value categories. 
- Most of these categorical fields having these unknown values have limited number of unknow values.
- However, poutcome (outcome of the previous campaign) has over 80% of its values as unknown, making it difficult to be used 

#### Outliers

- The balance field (average yearly bank balance) seems to be a bit un-reliable data source as there are a high percentage (7.8%) of zero avg balance and also high percentage (8.3%) of negative balance.
- The duration field (last contact duration in seconds) has some outliers in the higher end of the spread.
- Just like duration field, the campaign field (number of contacts performed in this campaign) has a large tail towards the higher end of the spread.
- pdays (number of days passed till last contacted) also has small number of data points stretched towards larger values.
- previous (contacts performed before this campaign) has one outlier value (valued at 275)

#### Skewed Fields

- poutcome
- previous
- pdays

#### Correlations between data features

- poutcome, previous and pdays are highly correlated because for 80% of the data-points they have unknown values
- age, job and education fields are highly correlated with each other
- contact, housing and month fields are highly correlated


# additional bank dataset

In [14]:
df = pd.read_csv("../dataset/bank-additional/bank-additional-full.csv", delimiter=';')
df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [15]:
profile = ProfileReport(df, title="Profiling Additional Bank dataset Report", correlations={
        "auto": {"calculate": False},
        "pearson": {"calculate": False},
        "spearman": {"calculate": False},
        "kendall": {"calculate": False},
        "phi_k": {"calculate": True},
        "cramers": {"calculate": True},
    })
os.makedirs("./../mount/eda_profile_reports", exist_ok=True)
profile.to_file("./../mount/eda_profile_reports/additional_bank_dataset.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]


### Notes

#### Duplicate entries
- Duplicate entries exists and need to be removed

#### Null values

- There are no case of Null value entries. 
- However some of the fields are having some unknown value categories. 
- Most of these categorical fields having these unknown values have limited number of unknow values.
- However, poutcome (outcome of the previous campaign) has over 80% of its values as unknown, making it difficult to be used 

#### Outliers

- The duration field (last contact duration in seconds) has some outliers in the higher end of the spread.
- Just like duration field, the campaign field (number of contacts performed in this campaign) has a large tail towards the higher end of the spread.
- pdays (number of days passed till last contacted) is set to 999 in the case where the contact is being made for the first time. Unlike -1 in the earlier dataset.
- pdays (number of days passed till last contacted) also has small number of data points stretched towards larger values.

#### Skewed/high unbalance Fields

- the default field (has credit in default?) is almost unusable field as most values are either no or unknown
- pdays

#### Correlations between data features

- poutcome, previous and pdays are highly correlated because for 80% of the data-points they have unknown values
- age, job and education fields are highly correlated with each other
- contact and month fields are highly correlated
- housing and load fields are highly correlated
